run/logging-manual/main.go

```diff
@@ -0,0 +1,102 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Sample Manual Logging shows how to leverage Cloud Run structured logging without a client library.
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"net/http"
+	"os"
+	"strings"
+
+	"cloud.google.com/go/compute/metadata"
+)
+
+var projectID string
+
+func main() {
+	projectID = os.Getenv("GOOGLE_CLOUD_PROJECT")
+	if projectID == "" {
+		projectID, _ = metadata.ProjectID()
+	}
+	if projectID == "" {
+		log.Println("Could not determine Google Cloud Project. Running without log correlation. For local use set the $GOOGLE_CLOUD_PROJECT environment variable.")
+	}
+
+	http.HandleFunc("/", indexHandler)
+
+	// Determine port for HTTP service.
+	port := os.Getenv("PORT")
+	if port == "" {
+		port = "8080"
+		fmt.Println("Defaulting to port", port)
+	}
+
+	// Start HTTP server.
+	fmt.Println("Listening on port", port)
+	log.Fatal(http.ListenAndServe(":"+port, nil))
+}
+
+// [START run_manual_logging_object]
+
+// Entry defines a log entry.
+type Entry struct {
+	Message  string `json:"message"`
+	Severity string `json:"severity,omitempty"`
+	Trace    string `json:"logging.googleapis.com/trace,omitempty"`
+
+	// Log viewer accesses 'component' as 'jsonPayload.component'.
+	Component string `json:"component,omitempty"`
+}
+
+// String renders an enty structure to the JSON format expected by Stackdriver.
+func (e Entry) String() string {
+	if e.Severity == "" {
+		e.Severity = "INFO"
+	}
+	out, _ := json.Marshal(e)
+	return string(out)
+}
+
+// [END run_manual_logging_object]
+// [START run_manual_logging]
+
+func indexHandler(w http.ResponseWriter, r *http.Request) {
+	// Uncomment and populate this variable in your code:
+	// projectID = "The project ID of your Cloud Run service"
+
+	// Derive the traceID associated with the current request.
+	var trace string
+	if projectID != "" {
+		traceHeader := r.Header.Get("X-Cloud-Trace-Context")
```

Comment:
I might put this into a separate function. I could see someone wanting to copy and paste just this part.

Reply:
The sample is in use at https://cloud.google.com/run/docs/logging#writing_structured_logs, are you suggesting a Go-specific change or a cross-language change? I'm open to a Go-specific change but would prefer to defer a larger change to follow-up.